### ISSUE TITLE
why train train bertwordpiece tokenizer with custom pre_tokenize&&normalize does not work??

### ISSUE THREAD

**Issue by labixiaoK** [huggingface/tokenizers#405]
hi, i want to train bertwordpiece tokenizer with custom normalizer and pre_tokenizer，it can train, but seems that custom module does not work, why this happened and how to solve it, the code is following: 
```
special_tokens = ["[PAD]", "[UNK]", "[MASK]", "[CLS]", "<S>", "</S>"] 
def _is_control(char):
    """Checks whether `char` is a control character."""
    # These are technically control characters but we count them as whitespace
    # characters.
    if char == "\t" or char == "\n" or char == "\r":
        return False
    cat = unicodedata.category(char)
    if cat.startswith("C"):
        return True
    return False


def _is_whitespace(char):
    """Checks whether `char` is a whitespace character."""
    # \t, \n, and \r are technically contorl characters but we treat them
    # as whitespace since they are generally considered as such.
    if char == " " or char == "\t" or char == "\n" or char == "\r":
        return True
    cat = unicodedata.category(char)
    if cat == "Zs":
        return True
    return False

class MybNormalizer():
    def __init__(self):
        pass

    def normalize(self, text):
        # logging.info('normalize in: ' + text)
        print('normalize in: ' + text)
        text = self._clean_text(text)
        text = text.lower()
        text = unicodedata.normalize('NFKC', text)
        # logging.info('normalize out: ' + text)
        print('normalize out: ' + text)
        return text

    def _clean_text(self, text):
        """Performs invalid character removal and whitespace cleanup on text."""
        text = text.strip()
        output = []
        for char in text:
            cp = ord(char)
            if cp == 0 or cp == 0xFFFD or _is_control(char):
                continue
            if _is_whitespace(char):
                output.append(" ")
            else:
                output.append(char)
        return "".join(output)


class MyPreprocessor():
    def __init__(self, my_dict_file='my-custom-vocab.txt'):
        if my_dict_file:
            jieba.load_userdict(my_dict_file)

    def pre_tokenize(self, sentence):
        # logging.info('pretokize in: ' + sentence)
        print('pretokize in: ' + sentence)
        """this function return tokenized version of the sentence, 
        it has to adjust the format of the output generated by jieba to be consistent with the one required by the tokenizers """
        # return [(token[0],(token[1],token[2])) for token in jieba.tokenize(sentence)]
        toks = jieba.cut(sentence, HMM=False)
        # logging.info('pretokize out: ' + str(toks))
        print('pretokize out: ' + str(toks))
        return [(tok, (0, 0)) for tok in toks]

    def decode(self, tokens):
        return ", ".join(tokens) 

   
def build_tokenizer():
    #全部false，自定义其中一些模块
    tokzer = BertWordPieceTokenizer(clean_text=False, handle_chinese_chars=False, lowercase=False, strip_accents=True)
    my_preprocessor = MyPreprocessor()
    tokzer.normalizer = MyNormalizer()
    tokzer.pre_tokenizer = pre_tokenizers.PreTokenizer.custom(my_preprocessor)
    tokzer.decoder = decoders.Decoder.custom(my_preprocessor)
    print('==============build tokzer: ', tokzer)    
    print('============== tokzer.normalizer: ', tokzer.normalizer)    
    print('============== tokzer.pretoker: ', tokzer.pre_tokenizer)    
    return tokzer

def main2(fls, fout):
    tokzer = build_tokenizer()
    tokzer.train(fls, vocab_size=32000, min_frequency=0, special_tokens=special_tokens, show_progress=True, limit_alphabet=7000, wordpieces_prefix="##")
    # And Save it
    tokzer.save(fout, pretty=True)

if __name__ =="__main__":
    fls = 'data/tmp.sents'.split(',')
    fout = 'mdl_wp/my_wp.mdl.test'
    main2(fls, fout)
```

when i run it, the logs in the `pre_tokenize `do not print, and the saved vocab has almost long sequences, seems not pre_tokenize, for example, part of vocab is as following:
```
      "##不愿意学呢": 1236,
      "因为我觉得如": 1237,
      "因为我自己对英语": 1238,
      "##的内容的话都是一些": 1239,
      "##什么时候有时间呢": 1240,
      "##什么时候有时间我给他": 1241,
      "00的单": 1242,

```

why this happend? 

lastly, my os is linux and python is 3.6 and tokenizer.__version__ is 0.8.1

**Comment by n1t0** [comment]
Hi @labixiaoK. Adding custom components is currently an alpha feature (even more like a proof-of-concept) and is not working properly. We are improving it, updating the API, and all of this should be available in the next release!

**Comment by labixiaoK** [comment]
> Hi @labixiaoK. Adding custom components is currently an alpha feature (even more like a proof-of-concept) and is not working properly. We are improving it, updating the API, and all of this should be available in the next release!

emmm....fine.  
another question, Does this only happen during training or also happen during using, because of ur example which use cumtom components when using tokenizer, 
https://github.com/huggingface/tokenizers/blob/master/bindings/python/examples/custom_pre_tokenizer.py  (sry for that I didn't try it myself)
btw, does the next release which u say mean v0.9.0 ? or 0.9.x ~

**Comment by n1t0** [comment]
Hi @labixiaoK. The new API for custom components has been released, and is available in any version > `0.9.0`. There are examples of custom components [here](https://github.com/huggingface/tokenizers/blob/master/bindings/python/examples/custom_components.py)

You can check [this answer](https://discuss.huggingface.co/t/how-to-add-additional-custom-pre-tokenization-processing/1637/2?u=anthony) on the forum for more information. Feel free to reopen if needed!
